### PR TITLE
Remove restriction for using printf for SYCL+CUDA

### PR DESCRIPTION
**Pull request by masterleinad**
We should be able to use `sycl::ONEAPI::experimental::printf` for `SYCL+CUDA` after https://github.com/intel/llvm/pull/4293.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
There were some problems with the `long long` overloads in `sycl::atomic_ref`, see https://github.com/intel/llvm/issues/4554.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

File: .jenkins [hunk 1]
```diff
@@ -44,10 +44,10 @@ pipeline {
                                 -DCMAKE_BUILD_TYPE=Release \
                                 -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
                                 -DCMAKE_CXX_COMPILER=clang++ \
-                                -DCMAKE_CXX_FLAGS="-Werror -Wno-unknown-cuda-version -Wno-gnu-zero-variadic-macro-arguments" \
+                                -DCMAKE_CXX_FLAGS="-Werror -Wno-unknown-cuda-version -Wno-unused-command-line-argument -Wno-gnu-zero-variadic-macro-arguments -Wno-linker-warnings" \
```
Comment (masterleinad) [inline .jenkins]:
```
error: linking module '/opt/sycl/lib/clang/14.0.0/../../clc/remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc': Linking two modules of different target triples: '/opt/sycl/lib/clang/14.0.0/../../clc/remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc' is 'nvptx64-unknown-nvidiacl' whereas '/tmp/kokkos/core/src/impl/Kokkos_CPUDiscovery.cpp' is 'nvptx64-nvidia-cuda'
 [-Werror,-Wlinker-warnings]
```

File: .jenkins [hunk 2]
```diff
@@ -44,7 +44,7 @@ pipeline {
                                 -DCMAKE_BUILD_TYPE=Release \
                                 -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
                                 -DCMAKE_CXX_COMPILER=clang++ \
-                                -DCMAKE_CXX_FLAGS="-Werror -Wno-unknown-cuda-version -Wno-gnu-zero-variadic-macro-arguments" \
+                                -DCMAKE_CXX_FLAGS="-Werror -Wno-gnu-zero-variadic-macro-arguments -Wno-linker-warnings" \
                                 -DKokkos_ARCH_VOLTA70=ON \
                                 -DKokkos_ENABLE_COMPILER_WARNINGS=ON \
                                 -DKokkos_ENABLE_DEPRECATED_CODE_3=ON \
```
Comment (masterleinad) [inline .jenkins:50]: There seems to be some problem with `impl/Kokkos_Timer.hpp` being included instead of `Kokkos_Timer.hpp` for `#include <Kokkos_Timer.hpp>`. I can't reproduce, though, and would want to move forward here independent of this weirdness.

Comment (masterleinad) [reply .jenkins]: See https://github.com/intel/llvm/issues/4226.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
OK. The `include` problem is fixed only in the nightly release from 09/10 which introduces more deprecations so I am in favor of just disabling deprecated code here and the 09/03 release.

File: .jenkins [hunk 3]
```diff
@@ -44,10 +44,10 @@ pipeline {
                                 -DCMAKE_BUILD_TYPE=Release \
                                 -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
                                 -DCMAKE_CXX_COMPILER=clang++ \
-                                -DCMAKE_CXX_FLAGS="-Werror -Wno-unknown-cuda-version -Wno-gnu-zero-variadic-macro-arguments" \
+                                -DCMAKE_CXX_FLAGS="-Werror -Wno-gnu-zero-variadic-macro-arguments -Wno-linker-warnings" \
                                 -DKokkos_ARCH_VOLTA70=ON \
                                 -DKokkos_ENABLE_COMPILER_WARNINGS=ON \
-                                -DKokkos_ENABLE_DEPRECATED_CODE_3=ON \
+                                -DKokkos_ENABLE_DEPRECATED_CODE_3=OFF \
```
Comment (dalg24) [inline .jenkins:50]: Why did you change?

File: .jenkins [hunk 4]
```diff
@@ -44,10 +44,10 @@ pipeline {
                                 -DCMAKE_BUILD_TYPE=Release \
                                 -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
                                 -DCMAKE_CXX_COMPILER=clang++ \
-                                -DCMAKE_CXX_FLAGS="-Werror -Wno-unknown-cuda-version -Wno-gnu-zero-variadic-macro-arguments" \
+                                -DCMAKE_CXX_FLAGS="-Werror -Wno-gnu-zero-variadic-macro-arguments -Wno-linker-warnings" \
```
Comment (dalg24) [inline .jenkins:47]: Why is it ok to drop `-Wno-unknown-cuda-version`

**Review by dalg24 — COMMENTED** [review]
Looks fine.  I just had two minor questions

File: .jenkins (hunk 3 above)
Comment (masterleinad) [reply .jenkins:50]: See https://github.com/kokkos/kokkos/pull/4244#discussion_r713284915 and https://github.com/kokkos/kokkos/pull/4244#issuecomment-925163562.

File: .jenkins (hunk 4 above)
Comment (masterleinad) [reply .jenkins:47]: Apparently, newer compiler versions don't complain about the CUDA version we are using anymore. :shrug: I tried to minimize the warnings we are ignoring here.

Comment (masterleinad) [reply .jenkins:47]: For the linker warning, see https://github.com/kokkos/kokkos/pull/4244#discussion_r713038908:
```
error: linking module '/opt/sycl/lib/clang/14.0.0/../../clc/remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc': Linking two modules of different target triples: '/opt/sycl/lib/clang/14.0.0/../../clc/remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc' is 'nvptx64-unknown-nvidiacl' whereas '/tmp/kokkos/core/src/impl/Kokkos_CPUDiscovery.cpp' is 'nvptx64-nvidia-cuda'
```

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
`HIP-ROCm-4.2-C++17` failing is clearly unrelated.

**Review by crtrott — CHANGES_REQUESTED** [review]
We have to limit the version enforcement for IntelLLVM to if SYCL is enabled.

**Review by crtrott — APPROVED** [review]
